plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/functions/tablechanges/TableChangesSplitSource.java

```diff
@@ -49,84 +51,95 @@
+                        CommitInfoEntry commitInfo = getOnlyElement(commitInfoEntries);
+
+                        List<ConnectorSplit> splits = new ArrayList<>();
+                        boolean containsCdcEntry = false;
+                        boolean containsRemoveEntry = false;
+                        for (DeltaLakeTransactionLogEntry entry : entries) {
+                            CdcEntry cdcEntry = entry.getCDC();
+                            if (cdcEntry != null) {
+                                containsCdcEntry = true;
+                                splits.add(mapToDeltaLakeTableChangesSplit(
+                                        commitInfo,
+                                        CDF_FILE,
+                                        cdcEntry.getSize(),
+                                        cdcEntry.getPath(),
+                                        cdcEntry.getCanonicalPartitionValues()));
+                            }
+                            if (entry.getRemove() != null && entry.getRemove().isDataChange()) {
+                                containsRemoveEntry = true;
+                            }
+                        }
+                        if (containsRemoveEntry && !containsCdcEntry) {
+                            throw new TrinoException(INVALID_FUNCTION_ARGUMENT, format("Change Data Feed is not enabled at version %d. Version contains 'remove' entries without 'cdc' entries", version));
+                        }
+                        if (!containsRemoveEntry) {
+                            for (DeltaLakeTransactionLogEntry entry : entries) {
+                                if (entry.getAdd() != null && entry.getAdd().isDataChange()) {
+                                    AddFileEntry addEntry = entry.getAdd();
+                                    splits.add(mapToDeltaLakeTableChangesSplit(
+                                            commitInfo,
+                                            DATA_FILE,
+                                            addEntry.getSize(),
+                                            addEntry.getPath(),
+                                            addEntry.getCanonicalPartitionValues()));
+                                }
+                            }
+                        }
+                        return splits.stream();
                     }
-                }
-            }
-            return CompletableFuture.completedFuture(new ConnectorSplitBatch(splits, processedVersion == tableReadVersion));
-        }
-        catch (IOException e) {
-            throw new TrinoException(DELTA_LAKE_FILESYSTEM_ERROR, "Failed to access table metadata", e);
+                    catch (IOException e) {
+                        throw new TrinoException(DELTA_LAKE_FILESYSTEM_ERROR, "Failed to access table metadata", e);
+                    }
+                });
+    }
+
+    @Override
+    public CompletableFuture<ConnectorSplitBatch> getNextBatch(int maxSize)
```

Comment:
Just for the understanding of the other reviewers: `maxSize` comes from `query.schedule-split-batch-size` setting.